frontend_v2/src/workspace/circuit.js

```diff
@@ -0,0 +1,226 @@
+			line += '----';
+		}
+		// Change of basis
+		if (form === 'cnot') {
+		    if (qubit.role === 'x') line += '-H--';
+		    if (qubit.role === 'z') line += '----';
+		} else if (form === 'univ') {
+		    line += '----';
+		}
+		// End
+		line += '----';
+		lines.push(line);
+
+		// Empty line with only vertical bars.
+		//      Q(__,__): ----
+		line = '              ';
+		line += '    ';
+		// Previous CNOTs and current one
+		for (let i = 0; i<idx+1; i++) {
+			line += ' |  ';
+		}
+		// Next CNOTs
+		for (let i = idx+1; i<data_qubits.length; i++) {
+			line += '    ';
+		}
+		line += '        '
+		lines.push(line);
+		idx += 1;
+	});
+	// Add line for the ancilla qubit.
+	let line = `${anc_qubit.name}: |0> `
+	if (form === 'cnot') {
+		line += '----'; // During the change of basis
+	} else if (form === 'univ') {
+		line += '-H--';
+	}
+	for (let i = 0; i<data_qubits.length; i++) {
+		if (form === 'cnot') {
+			line += '-X--';
+		} else if (form === 'univ') {
+			line += '-*--';
+		}
+	}
+	if (form === 'cnot') {
+		line += '---- D~ '
+	} else if (form === 'univ') {
+		line += '-H-- D~ ';
+	}
+	lines.push(line);
+	// Add time ruler.
+	if (with_time) {
+		//      Q(__,__): 
+		line = '          ';
+		let line2 = 'time ruler';
+		for (let i = 0; i<data_qubits.length+4; i++) {
+			line += '___ ';
+			line2 += ' ' + i + '  ';
+		}
+		lines.push(line);
+		lines.push(line2);
+	}
+	// Create the message
+	let art = '';
+	lines.slice(0,-1).forEach(line => {
+		art = art + line + '\n';
+	});
+	art += lines[lines.length-1];
+	return art;
+}
+
+
+
+/**
+ * Circuit class
+ * @extends Container
+ * @constructor
+ * @param {array} qubits - The qubits involved in the circuit
+ * @param {number} x - The x position of the circuit graphics
+ * @param {number} y - The y position of the circuit graphics
+ * @param {number} color - Color filling of the circuit
+ */
+export default class Circuit extends Container {
+	constructor(qubits, x, y, color = 'purple', message = '') {
+		super();
+		// Color properties (as static fields).
+		// Associated to the role played by the qubit.
+		Circuit.color_background = color
+		Circuit.color = 'white'
+		// UI properties
+		this.globalX = x;
+		this.globalY = y;
+		this.isCompatible = true;
+		//this.anc_qubit;
+		this.data_qubits = [];
+		this._confirmFormat(qubits);
+		this.art = this._createCircuit(message);
+		this.box = this._createBackground();
+		// Add the text object to the stage
+		this.addChild(this.box);
+		this.addChild(this.art);
+	}
+
+	/**
+	 * Confirm format of the plaquette, constraining the type of circuit
+	 * - there is a unique ancilla qubit
+	 * - all other qubits are associated with either X- or Z-stabilizers
+	 * - all data qubits are assumed connected to the ancilla qubit
+	 *   (i.e. they can be involved in 2q gates)
+	 */
+	_confirmFormat(qubits) {
+		this.isCompatible = true;
+		let numAncillas = 0;
+        qubits.forEach(qubit => {
+            if (qubit.role === 'a') {
```

Comment:
To make it totally clear, create a set of symbols like so:
```
QUBIT_ROLES = {
  ANCILLA: Symbol('a'),
  MEASURE: ...
  ...
}
```

And update the references

Reply:
Is the use of "all capitalized characters" the recommended style for this sort of dictionary?

Reply:
Since this PR is already merged, I will open another one with this change (and other changes suggested in the review).